include/oneapi/dpl/internal/random_impl/geometric_distribution.h

```diff
@@ -0,0 +1,193 @@
+namespace oneapi
+{
+namespace dpl
+{
+template <class _IntType = int>
+class geometric_distribution
+{
+  public:
+    // Distribution types
+    using result_type = _IntType;
+    using scalar_type = internal::element_type_t<_IntType>;
+
+    struct param_type
+    {
+        param_type() : param_type(0.5) {}
+        param_type(double __p) : p(__p) {}
+        double p;
+    };
+
+    // Constructors
+    geometric_distribution() : geometric_distribution(0.5) {}
+    explicit geometric_distribution(double __p) : p_(__p) {}
+    explicit geometric_distribution(const param_type& __params) : p_(__params.p) {}
+
+    // Reset function
+    void
+    reset()
+    {
+    }
+
+    // Property functions
+    double
+    p() const
+    {
+        return p_;
+    }
+
+    param_type
+    param() const
+    {
+        return param_type(p_);
+    }
+
+    void
+    param(const param_type& __param)
+    {
+        p_ = __param.p;
+    }
+
+    scalar_type
+    min() const
+    {
+        return scalar_type{};
+    }
+
+    scalar_type
+    max() const
+    {
+        return std::numeric_limits<scalar_type>::max();
+    }
+
+    // Generate functions
+    template <class _Engine>
+    result_type
+    operator()(_Engine& __engine)
+    {
+        return operator()<_Engine>(__engine, param_type(p_));
+    }
+
+    template <class _Engine>
+    result_type
+    operator()(_Engine& __engine, const param_type& __params)
+    {
+        return generate<size_of_type_, _Engine>(__engine, __params);
+    }
+
+    template <class _Engine>
+    result_type
+    operator()(_Engine& __engine, unsigned int __random_nums)
+    {
+        return operator()<_Engine>(__engine, param_type(p_), __random_nums);
+    }
+
+    template <class _Engine>
+    result_type
+    operator()(_Engine& __engine, const param_type& __params, unsigned int __random_nums)
+    {
+        return result_portion_internal<size_of_type_, _Engine>(__engine, __params, __random_nums);
+    }
+
+  private:
+    // Size of type
+    static constexpr int size_of_type_ = internal::type_traits_t<result_type>::num_elems;
+
+    // Static asserts
+    static_assert(::std::is_integral<scalar_type>::value,
+                  "oneapi::dpl::geometric_distribution. Error: unsupported data type");
+
+    // Distribution parameters
+    double p_;
+
+    // Implementation for generate function
+    template <int _Ndistr, class _Engine>
+    typename ::std::enable_if<(_Ndistr != 0), result_type>::type
+    generate(_Engine& __engine, const param_type& __params)
+    {
+        return generate_vec<_Ndistr, _Engine>(__engine, __params);
+    }
+
+    // Specialization of the scalar generation
+    template <int _Ndistr, class _Engine>
+    typename ::std::enable_if<(_Ndistr == 0), result_type>::type
+    generate(_Engine& __engine, const param_type& __params)
+    {
+        oneapi::dpl::uniform_real_distribution<double> __u;
+        return sycl::floor(sycl::log(1.0 - __u(__engine)) / sycl::log(1.0 - __params.p));

```

Comment:
Does it make sense to precompute sycl::log(1.0 - __params.p) and save as a private parameter? It can be applicable when operator() is called without additional parameters

Reply:
This will not lead to much optimization, because our main use case involves creating a new generator in the kernel.